pontoon/static/js/tags/admin/index.js

```diff
@@ -0,0 +1,80 @@
+
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import TagResourceManager from './manager';
+
+
+export class TagResourcesButton extends React.Component {
+
+    constructor (props) {
+        super(props);
+        this.handleClick = this.handleClick.bind(this);
+        this.state = {open: false};
+    }
+
+    handleClick (evt) {
+        evt.preventDefault();
+        const {open} = this.state;
+        this.setState({open: !open});
+    }
+
+    renderButton () {
+        const {open} = this.state;
+        let message = "Manage resources for this tag";
+        if (open) {
+            message = "Hide the resource manager for this tag";
+        }
+        return (
+              <button onClick={this.handleClick}>
+                {message}
+              </button>);
+    }
+
+    renderResourceManager () {
+        const {open} = this.state;
+        if (!open) {
+            return '';
+        }
+        return (
+            <TagResourceManager
+               tag={this.props.tag}
+               api={this.props.api}
+               project={this.props.project} />);
+    }
+
+    render () {
+        return (
+            <div>
+              {this.renderButton()}
+              {this.renderResourceManager()}
+            </div>);
+    }
+}
+
+
+export class TagResourcesLoader {
+
+    get tagResourceNodes () {
+        return Array.from(document.querySelectorAll('div.js-tag-resources'));
+    }
+
+    addButton (node) {
+        ReactDOM.render(
+            <TagResourcesButton
+               api={node.dataset.api}
+               tag={node.dataset.tag}
+               project={node.dataset.project} />,
+            node);
+    }
+
+    addButtons () {
+        this.tagResourceNodes.forEach(this.addButton);
+    }
+}
```

Comment:
> I don't understand why this is a class.

I can very much imagine this being overridden as a generic loader - i would defo tend towards classes over functions - even with js

Reply:
I get your point, but I'll argue that this is increasing complexity for the sake of maybe making something simpler in the future. My worry is that we might never actually need a loader, and we thus simply have code that's more complex for no good reason. I've seen this happen on Socorro, and make some parts of the code base a nightmare to maintain, because the code was way too generic. I'd love to avoid this happening on Pontoon. :)

Reply:
Im going to have to disagree here. Well structured classes, with good names for methods and properties are more readable, maintainable, extensible and testable than procedural code with closure, im struggling to see how you think the example you gave is easier to understand

Reply:
As a general comment, if there's code in need of more detail to understand, let's add comments, and not more code.

Also, in this particular example, there's nothing to test in each method, aside of the Array.from and NodeList implementations in browsers. the only way to test that this code actually does what it should is, at this level, an integration test of the function or class.

And please, no generic base class, in a separate file, even, for use-cases like this.

Nit on Adrian's code, there's an optional second argument to `Array.from`, which comes in handy in cases like this. You can pass in a function that'll be called for each item before the Array is constructed. One-liner.

Reply:
> As a general comment, if there's code in need of more detail to understand, let's add comments, and not more code

For me, its not so much about adding more/less code, its more about preferring an object-oriented style and use of data structures over functional/closure style.

> Also, in this particular example, there's nothing to test in each method

I disagree. We want to know that the selectors being used in the js widget match the ones being used when generating the templates:

https://github.com/phlax/pontoon/blob/658bd64a17eee5d779c6a226070184974e2d993b/tests/js/tags/admin.js#L486

There is literally nothing else that ties these 2 disparate parts of the code together.

> And please, no generic base class, in a separate file, even, for use-cases like this.

None were used. My point was more that the thing that has sold me on React, is the ease with which i can build components, see common patterns, factor out the useful stuff, rinse and repeat.

Reply:
There are three things that bother me, and make me favor "my way". First, I believe code complexity is highly coupled to the number of lines of code. Less lines of code, means less things to read, which, most of the time, means code is simpler to understand. 

Second, in this specific case, using a class means that to understand how the algorithm plays out, you have to jump from method to method. With the function, it goes from top to bottom, you just have to read it straight, no "goto". It seems a lot simpler to me, especially with such a small function. Why do complicated if you can do simple? (as we often say in French, not sure that is an actual idiom in English... :P ) That leads me to believe this is a case of premature optimization. You write code anticipating use cases that do not exist yet, and might never exist. But I've made that point already.

My last point goes deeper, it's on using OOP. I've grown very wary of OOP, because it leads to some antipatterns that can be very difficult to deal with, especially in big code bases. The more I code, and the more I use the React + redux environment, the more I think OOP should be the exception. I do use classes when it makes sense, for stateful components for example. But I avoid inheritance as much as possible, favoring composition instead. 

The best resources I have on this point are these two articles by Eric Elliott, on what he calls the "Two pillars of JS": https://medium.com/javascript-scene/the-two-pillars-of-javascript-ee6f3281e7f3 and https://medium.com/javascript-scene/the-two-pillars-of-javascript-pt-2-functional-programming-a63aa53a41a4

Reply:
hmm - i definitely get that functional programming has classically been the style of javascript - and could find many articles bestowing its virtue (and tbh critiquing it).

imhe, its difficult to test, and leads to heavily nested code in which its very difficult to keep track of local vars. I have personally debugged a lot of javascript code in which these vars were for one reason or another not being garbage collected. Found using a profiler, because the code was so hard to follow.

I think modern javascript has to a large extent addressed these issues by:

- introducing a more familiar OOP pattern (i get that this is "syntactic sugar" but its powerful all the same)
- introducing (like python) `await` syntax that lets you not have to keep embedding functionality in ever more nested code

just my 2p, but i would like to take adavantage of these advances

Reply:
After studying the tests, I've opened an email thread about the goals of our testing. In this particular case, yes, the tests only test the additional lines of code around the actual for loop, and validate that they don't break. I'd make that point on the actual tests, but github refuses to load that part of the diff for me.

React-newbie question: where are the <div class="js-tag-resources"> elements actually created? Didn't find that.

Reply:
@Pike in the following PR here - #802

Reply:
more specifically here https://github.com/mozilla/pontoon/pull/802/files#diff-d7b0530da8eb2ab60bc18f6d157c4a73R24